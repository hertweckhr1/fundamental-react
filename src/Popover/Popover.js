import { chain } from 'chain-function';
import Popper from '../utils/_Popper';
import { POPPER_PLACEMENTS } from '../utils/constants';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

class Popover extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isExpanded: false
        };
    }

    triggerBody = () => {
        if (!this.props.disabled) {
            this.setState(prevState => ({
                isExpanded: !prevState.isExpanded
            }));
        }
    };

    handleOutsideClick = () => {
        if (this.state.isExpanded) {
            this.setState({
                isExpanded: false
            });
        }
    };

    render() {
        const {
            disabled,
            noArrow,
            control,
            body,
            className,
            placement,
            ...rest
        } = this.props;

        let onClickFunctions = this.triggerBody;
        if (control.props.onClick) {
            onClickFunctions = chain(this.triggerBody, control.props.onClick);
        }

        const referenceComponent = React.cloneElement(control, {
            onClick: onClickFunctions
        });

        return (
            <Popper
                noArrow={noArrow}
                onClickOutside={this.handleOutsideClick}
                onEscapeKey={this.handleOutsideClick}
                popperClassName={className}
                popperPlacement={placement}
                popperProps={rest}
                referenceClassName='fd-popover'
                referenceComponent={referenceComponent}
                show={this.state.isExpanded && !disabled}>
                {body}
            </Popper>
        );
    }
}

Popover.displayName = 'Popover';

Popover.propTypes = {
    body: PropTypes.node.isRequired,
    control: PropTypes.node.isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    noArrow: PropTypes.bool,
    placement: PropTypes.oneOf(POPPER_PLACEMENTS)
};

Popover.propDescriptions = {
    body: 'Node(s) to render in the overlay.',
    control: 'Node to render as the reference element (that the `body` will be placed in relation to).',
    noArrow: 'Set to **true** to render a popover without an arrow.',
    placement: 'Initial position of the `body` (overlay) related to the `control`.'
};

export default Popover;
