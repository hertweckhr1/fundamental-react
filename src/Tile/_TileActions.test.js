import { Button } from '../';
import Menu from '../Menu/Menu';
import { mount } from 'enzyme';
import Popover from '../Popover/Popover';
import React from 'react';
import renderer from 'react-test-renderer';
import TileActions from './_TileActions';

describe('<TileActions />', () => {
    const actionTile = (
        <TileActions className='yellow'>
            <Popover
                body={
                    <Menu>
                        <Menu.List>
                            <Menu.Item url='/'>Option 1</Menu.Item>
                            <Menu.Item url='/'>Option 2</Menu.Item>
                            <Menu.Item url='/'>Option 3</Menu.Item>
                            <Menu.Item url='/'>Option 4</Menu.Item>
                        </Menu.List>
                    </Menu>
                }
                control={<Button glyph='vertical-grip' type='standard' />} />
        </TileActions>
    );

    test('create tile actions component', () => {
        // action tile
        let component = renderer.create(actionTile);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the TileActions component', () => {
            const element = mount(<TileActions data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });
});
