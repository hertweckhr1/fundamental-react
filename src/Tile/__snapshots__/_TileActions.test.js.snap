// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TileActions /> create tile actions component 1`] = `
<div
  className="fd-tile__actions yellow"
>
  <div
    className="fd-popover"
  >
    <div
      aria-expanded={false}
      className="fd-popover__control"
      onClick={[Function]}
    >
      <button
        className="fd-button fd-button--standard sap-icon--vertical-grip"
      />
    </div>
    <div
      aria-hidden={true}
      className="fd-popover__body"
    >
      <nav
        className="fd-menu"
      >
        <ul
          className="fd-menu__list"
        >
          <li>
            <a
              className="fd-menu__item"
              href="/"
            >
              Option 1
            </a>
          </li>
          <li>
            <a
              className="fd-menu__item"
              href="/"
            >
              Option 2
            </a>
          </li>
          <li>
            <a
              className="fd-menu__item"
              href="/"
            >
              Option 3
            </a>
          </li>
          <li>
            <a
              className="fd-menu__item"
              href="/"
            >
              Option 4
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
